### PR TITLE
cmake: add legacy-option-headers' sources with INTERFACE

### PR DESCRIPTION
**Pull request by tchaikov**
to address the failure when generating the building system like:

```
CMake Error at src/common/options/CMakeLists.txt:108 (target_sources):
  target_sources may only set INTERFACE properties on INTERFACE targets
```

let's add `legacy-option-headers`'s source files with `INTERFACE`.
also, in order to workaround
https://gitlab.kitware.com/cmake/cmake/-/issues/18399, we have to set
the GENERATED property of the generated header file in src/ so that it's
visible to all cmake script where the dependency should be populated.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
pushed to ceph-ci, and built at https://shaman.ceph.com/builds/ceph/wip-pr-58078-kefu/4630b638d3ea6ec43a3b9634c29a930787f32041/

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -629,6 +629,17 @@ target_link_libraries(rados_snap_set_diff_obj
...
 option(WITH_LIBRADOSSTRIPER "build with libradosstriper support" ON)
 
+if(CMAKE_VERSION VERSION_LESS "3.20")
+  # see src/common/options/CMakeLists.txt
+  set(all_options
+    global cephfs-mirror crimson mgr mds mds-client mon osd rbd rbd-mirror immutable-object-cache ceph-exporter)
+  foreach(name ${all_options})
+    set_property(SOURCE
+      ${PROJECT_BINARY_DIR}/include/${name}_legacy_options.h
+      PROPERTY GENERATED TRUE)
```
Comment (idryomov) [inline src/CMakeLists.txt:639]: This still fails for me with a slew of errors like in https://gitlab.kitware.com/cmake/cmake/-/issues/18399:
```
-- Configuring done
CMake Error at cmake/modules/BuildBoost.cmake:307 (_add_executable):
  Cannot find source file:

    /home/dis/src/ceph/build/include/rgw_legacy_options.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx
Call Stack (most recent call first):
  src/CMakeLists.txt:720 (add_executable)


CMake Error at cmake/modules/BuildBoost.cmake:299 (_add_library):
  Cannot find source file:

    /home/dis/src/ceph/build/include/global_legacy_options.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx
Call Stack (most recent call first):
  src/auth/CMakeLists.txt:23 (add_library)


CMake Error at cmake/modules/BuildBoost.cmake:299 (_add_library):
  Cannot find source file:

    /home/dis/src/ceph/build/include/global_legacy_options.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx
Call Stack (most recent call first):
  src/common/CMakeLists.txt:184 (add_library)

[...]

-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```
All of them are coming from `cmake/modules/BuildBoost.cmake`, line 299 or 307:
```
297 # override add_library() to add Boost headers dependency
298 function(add_library target)                 
299   _add_library(${target} ${ARGN})      
300   # can't add dependencies to aliases or imported libraries
301   if (NOT ";${ARGN};" MATCHES ";(ALIAS|IMPORTED);")           
302     maybe_add_boost_dep(${target})     
303   endif()
304 endfunction()                      
305        
306 function(add_executable target)
307   _add_executable(${target} ${ARGN})                    
308   maybe_add_boost_dep(${target})                          
309 endfunction()
```

**Comment by tchaikov** [comment]
> we might just bump the minimum cmake version now that we've dropped support for focal

i really wish we could

**Comment by idryomov** [comment]
> > we might just bump the minimum cmake version now that we've dropped support for focal
> 
> i really wish we could when looking at the ugly workaround. but it's probably up to @idryomov ?

It's not up to me ;)  IMO it's very useful to be able to build all supported releases in the same environment without juggling several containers or similar.  It has been working for a long time and I'm routinely taking advantage of that, especially when needing to test backwards compatibility with different versions of daemons and all client libraries available alongside.

If the workaround is bearable (i.e. just ugly in the sense of needing to stick a line in every `CMakeLists.txt` as opposed to being technically fragile), my vote would to stick with it.

**Comment by cbodley** [comment]
> IMO it's very useful to be able to build all supported releases in the same environment without juggling several containers or similar.

ubuntu jammy is supported from quincy through main. jammy's cmake version is 3.22.1

**Comment by idryomov** [comment]
> If the workaround is bearable (i.e. just ugly in the sense of needing to stick a line in every `CMakeLists.txt` as opposed to being technically fragile), my vote would to stick with it.

@tchaikov FWIW the current version still fails for me with the same errors.

**Comment by idryomov** [comment]
> > IMO it's very useful to be able to build all supported releases in the same environment without juggling several containers or similar.
> 
> ubuntu jammy is supported from quincy through main. jammy's cmake version is 3.22.1

Heh, I guess should have said "all recent releases".  I often find myself needing to build something based on pacific (for downstream purposes or otherwise) and I expect this to continue for quite some time.  The same was the case with nautilus which was a major release in the sense of the number of deployments.

If I'm the only one doing this, I definitely won't stand in the way of bumping the required version.

**Comment by idryomov** [comment]
> If I'm the only one doing this, I definitely won't stand in the way of bumping the required version.

I will say that the build system doesn't strike me as the component where the version requirement needs to be advanced as soon as possible, but I also haven't been involved with our CMake scripts beyond occasional tweaks, so I might very well be misjudging the issues.

**Comment by tchaikov** [comment]
> > If the workaround is bearable (i.e. just ugly in the sense of needing to stick a line in every `CMakeLists.txt` as opposed to being technically fragile), my vote would to stick with it.
> 
> @tchaikov FWIW the current version still fails for me with the same errors.

i tested locally with 
```bash
$ cmake -GNinja -Bbuild.cmake -DCMAKE_BUILD_MODE=Debug -DWITH_ASAN=ON -DWITH_MANPAGE=OFF -DWITH_LIBCEPHSQLITE=OFF -DWITH_OPENLDAP=OFF -DWITH_FUSE=OFF -DWITH_RBD=OFF -DWITH_KRBD=OFF -DWITH_RADOSGW=OFF -DWITH_BLKIN=OFF -DWITH_LTTNG=OFF -DWITH_BABELTRACE=OFF -DWITH_LIBCEPHFS=OFF -DWITH_SYSTEM_BOOST=ON -DWITH_CEPHFS=OFF
```
as i am disabling quite a few features due to the missing dependencies. there are good chances that i missed some directories, would you please share the error messages, so i can fix the failure accordingly?

**Comment by tchaikov** [comment]
> > > we might just bump the minimum cmake version now that we've dropped support for focal
> > 
> > 
> > i really wish we could when looking at the ugly workaround. but it's probably up to @idryomov ?
> 
> It's not up to me ;) IMO it's very useful to be able to build all supported releases in the same environment without juggling several containers or similar. It has been working for a long time and I'm routinely taking advantage of that, especially when needing to test backwards compatibility with different versions of daemons and all client libraries available alongside.

if i am reading your reply correctly, does it imply that: if ceph x.y.z compiles with CMake 3.22 and up, you would accept the change?

> 
> If the workaround is bearable (i.e. just ugly in the sense of needing to stick a line in every `CMakeLists.txt` as opposed to being technically fragile), my vote would to stick with it.

but it hurts maintainability. "ugly" is but a subjective way to put this. if i had the privilege to vote, i am inclined to avoid this.

**Comment by idryomov** [comment]
> as i am disabling quite a few features due to the missing dependencies. there are good chances that i missed some directories, would you please share the error messages, so i can fix the failure accordingly?

The error messages are the same as reported in https://github.com/ceph/ceph/pull/58078#discussion_r1642715596.  I think the reason that you aren't seeing it is that you are passing `-DWITH_SYSTEM_BOOST=ON` -- I think it defaults to OFF because Ubuntu Jammy doesn't have Boost 1.82 which is required by main.

Are you installing ceph-libboost packages to make `-DWITH_SYSTEM_BOOST=ON` work?

**Comment by idryomov** [comment]
> if i am reading your reply correctly, does it imply that: if ceph x.y.z compiles with CMake 3.22 and up, you would accept the change?

No.  My wish (it's not a formal requirement of the project, so you don't need _me_ to accept anything) is rather vague, but in a nutshell it would be something along the lines of "let things continue to be able to build and run on older distros for as long as possible" or, put another way, "don't bump version version requirements of both build and runtime dependencies on core/non-optional components unless it's necessary".  Since CMake is the gate to everything, its version is particularly sensitive.

> where x belongs to a set of `{17, 18}`? see https://docs.ceph.com/en/latest/releases/

As mentioned in https://github.com/ceph/ceph/pull/58078#issuecomment-2174032918, for me the set extends beyond officially supported releases.  I have been enjoying the ability to build and run core daemons and libraries on all branches starting with nautilus all the way up to main alongside on the same node (in my case Ubuntu Focal with just g++ upgraded for C++20).  I realize that it's something that can't work forever, but it's sad to see it break because of CMake as opposed to something more tangible.

**Comment by tchaikov** [comment]
> > as i am disabling quite a few features due to the missing dependencies. there are good chances that i missed some directories, would you please share the error messages, so i can fix the failure accordingly?
> 
> The error messages are the same as reported in [#58078 (comment)](https://github.com/ceph/ceph/pull/58078#discussion_r1642715596). I think the reason that you aren't seeing it is that you are passing `-DWITH_SYSTEM_BOOST=ON` -- I think it defaults to OFF because Ubuntu Jammy doesn't have Boost 1.82 which is required by main.

i see.

> 
> Are you installing ceph-libboost packages to make `-DWITH_SYSTEM_BOOST=ON` work?

nope. i was using fedora 40. IIRC, it's able to provide the necessary dependencies.

**Comment by tchaikov** [comment]
> > if i am reading your reply correctly, does it imply that: if ceph x.y.z compiles with CMake 3.22 and up, you would accept the change?
> 
> No. My wish (it's not a formal requirement of the project, so you don't need _me_ to accept anything) is rather vague, but in a nutshell it would be something along the lines of "let things continue to be able to build and run on older distros for as long as possible" or, put another way, "don't bump version requirements of build or runtime dependencies on core/non-optional components unless necessary". Since CMake is the gate to everything, its version is particularly sensitive.
> 
> > where x belongs to a set of `{17, 18}`? see https://docs.ceph.com/en/latest/releases/
> 
> As mentioned in [#58078 (comment)](https://github.com/ceph/ceph/pull/58078#issuecomment-2174032918), for me the set extends beyond officially supported releases. I have been enjoying the ability to build and run core daemons and libraries on all branches starting with nautilus all the way up to main alongside on the same node (in my case Ubuntu Focal with just g++ upgraded for C++20). I realize that it's something that can't work forever, but it's sad to see it break because of CMake as opposed to something more tangible.

i see. but you can always use ubuntu/focal as long as you want, a recent release CMake can be installed locally. we upgraded the supported building platform from xenial to bionic and bumped up the minimal required CMake version to 3.10. and later we dropped the support of bionic, and bumped it to 3.16. i am sure, this will happen sooner or later either by me or by somebody else who maintains and cares about the CMake building system.

**Comment by cbodley** [comment]
@tchaikov i pushed a draft https://github.com/ceph/ceph/pull/58238 which replaces `target_sources()` with `add_dependencies()` to track `legacy-option-headers`' dependencies on these generated headers. i'm hoping that works around https://gitlab.kitware.com/cmake/cmake/-/issues/18399 to avoid the need to bump cmake's minimum version

it seems to work with my cmake 3.28.2 but i haven't tested with older cmakes

**Pull request closed without merging**